### ISSUE TITLE
how to train yolov5 to detect small objects

### ISSUE THREAD

**Issue by Grenaf** [obss/sahi#201]
Hello author！I have some questions about how to train yolov5 when i use sahi in large scale images.
If my data set image pixels are 1920x1080 size. What image pixel size should I use to train the network?
# Training phase
Assume that the input-size of the initial training model is 640x640 (can be changed). Method 1: The original 1920x1280 pixel picture is sent to the model training (the original image information will be lost after resize) Method 2: The original image slices are sent to the yolov5 network for training.
# Detection stage
 Should the sliding window be consistent with the input-size of the model?
Reference website: https://github.com/ultralytics/yolov5/issues/46

# Extra question
When I used the colab demo of yolov5+sahi, I modified the following parameters, and some unexpected situations occurred and I was a little confused.Some extra bounding boxes, some small targets under a 256x256 sliding window are missed under a 640x640 sliding window
![image](https://user-images.githubusercontent.com/73221711/130077885-267989da-e8fb-477a-9d75-f397b17882c7.png)

# Question summary:
1. I would like to ask the author how the input-size is set in the training phase of your demo. By the way, can you share the demo highway vehicle data set. 
2. What size picture should the user prepare before training the network. 
3. How to set the size of the sliding window in the detection stage to be appropriate

**Comment by fcakyon** [comment]
@Grenaf Thanks a lot for you interest and valuable comments!

1. Demo uses coco pretrained weight from yolov5 and mmdetection repos. There is no sliced fine-tuning included in the demo. However, sliced fine-tuning increases the performance.
2. Size of the slices are determined by your object and image sizes. If you have 10-20pixel objects with 1000*1000 images 512*512 or 256*256 image sizes could be great choice. If you have larger objects as 200*200, 640*640 slice size would be better.
3. You can prepare a validation set and evaluate the performance of your val set with different slicesizes to determine the optimum slice size.

**Comment by Grenaf** [comment]
Thank you sir! I will try your methods. 
# new problem
- But with a new problem, yolov5 uses mosaic image enhancement technology to solve the problem of small target recognition . If I use the default mosaic function(small target pixels will lose part of it), will all the pictures of the network training be the pictures after the mosaic function? Or the picture that enters the network training is the original picture + mosaic picture.Should I use the mosaic function?
- Is the sliding window size the same as the detection network input size?
parser.add_argument('--imgsz', '--img', '--img-size', type=int, default=640, help='inference size (pixels)')（yolov5‘s detect.py）
![image](https://user-images.githubusercontent.com/73221711/130209294-ec5f1c1b-da4c-474e-84fb-aa7f0035f129.png)
# The source of the problem
- https://github.com/ultralytics/yolov5/issues/46
![image](https://user-images.githubusercontent.com/73221711/130207539-9cfb9564-5b00-4f73-83fb-fbcb0ebbb11e.png)
- https://github.com/ultralytics/yolov5/issues/3709
![image](https://user-images.githubusercontent.com/73221711/130208040-899d06a9-46b6-4a7f-b1b4-eee1aa8c0340.png)
# yolov5's new function
- https://github.com/ultralytics/yolov5/issues/3942 
I find a new feature of yolov5 to share with you!

**Comment by fcakyon** [comment]
@Grenaf you can set the yolov5 inference image size with `image_size` parameter of `get_sliced_prediction` function: https://github.com/obss/sahi/blob/d5e5e7d77b2fc4eb35b550deb418b51f46615fd4/sahi/predict.py#L108, `get_prediction` function: https://github.com/obss/sahi/blob/d5e5e7d77b2fc4eb35b550deb418b51f46615fd4/sahi/predict.py#L27, or `predict` function: https://github.com/obss/sahi/blob/d5e5e7d77b2fc4eb35b550deb418b51f46615fd4/sahi/predict.py#L279.

When you dont provide `image_size` parameter, default value for Yolov5 is `640`.

**Comment by fcakyon** [comment]
@Grenaf there are 2 alternatives to sliced fine-tuning:
- slice the whole dataset and use the sliced images/annotations during training: https://github.com/obss/sahi/blob/main/scripts/slice_coco.py
- apply randomcrop augmentation with albumentations as you mentioned above.

If you have any further questions feel free to ask!

**Comment by fcakyon** [comment]
@Grenaf by default, yolov5 uses only mosaics during the training and that is where the power of yolo comes from.

**Comment by Grenaf** [comment]
Thank you for your patience, some questions still puzzle me.
# questions
-you say yolov5 uses only mosaics during the training and that is where the power of yolo comes from if that means training network will just use mosaic’s images（Splicing and resize from original images） ，original images will not be trained in training network？
-What is the relationship between inference size and slice height and slice weight?

**Comment by fcakyon** [comment]
Your first question is not sahi related, you would get more indepth response from the original yolov5 author :)

**Comment by Grenaf** [comment]
OK!The second answer is so funny！If I set slice size 640 and the inference size is 640，so slices will not be resized？Your postprocess is so funny! I used to see some familiar way just like YOLT，I will continue to use sahi！Thanks！

**Comment by fcakyon** [comment]
@Grenaf you are correct, when slice and input size is same, images will not be scaled :)

You can consider `sahi` as the improved version of `yolt` with extra features such as:
- fiftyone support: https://github.com/obss/sahi/releases/tag/0.5.1
- coco utilities: https://github.com/obss/sahi/blob/main/docs/COCO.md
- mmdetection support: https://colab.research.google.com/github/obss/sahi/blob/main/demo/inference_for_mmdetection.ipynb
- coco error analysis scripts: https://github.com/obss/sahi/blob/main/docs/SCRIPTS.md#coco_error_analysispy-script-usage
